mux/mux.go

```diff
@@ -3,104 +3,87 @@ package mux
 	}
-	if mux.httpServer == nil && mux.grpcServer == nil {
+	if mux.httpHandler == nil && mux.grpcServer == nil {
 		return errors.New("at-least one of http & grpc server must be set")
 	}
-
 	return mux.Serve(ctx, listenAddr)
 }
 
-type cmuxWrapper struct {
-	httpServer  *http.Server
+type protocolMuxServer struct {
+	httpHandler http.Handler
 	grpcServer  *grpc.Server
 	gracePeriod time.Duration
 }
 
-func (cmw *cmuxWrapper) Serve(baseCtx context.Context, addr string) error {
+func (pmux *protocolMuxServer) Serve(baseCtx context.Context, addr string) error {
 	ctx, cancel := context.WithCancel(baseCtx)
 	defer cancel()
 
-	l, err := net.Listen("tcp", addr)
-	if err != nil {
-		return err
+	http1Server := &http.Server{
+		Addr:    addr,
+		Handler: h2c.NewHandler(pmux.muxedHandler(), &http2.Server{}),
 	}
-	defer l.Close()
-
-	m := cmux.New(l)
-	defer m.Close()
-
-	grpcL := m.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
-	httpL := m.Match(cmux.Any())
-
-	errorCh := make(chan error, 3)
-	serveOnListener(httpL, cmw.httpServer, errorCh)
-	serveOnListener(grpcL, cmw.grpcServer, errorCh)
 
 	go func() {
-		if err := m.Serve(); err != nil {
-			errorCh <- err
+		err := http1Server.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Printf("[ERROR] server exited with error: %v", err)
+
+			// force-cancel the context so that graceful shutdown sequence exits as well.
+			cancel()
 		}
 	}()
 
-	select {
-	case <-ctx.Done():
-		return cmw.shutDown()
+	<-ctx.Done()
 
-	case e := <-errorCh:
-		_ = cmw.shutDown()
-		return e
-	}
-}
-
-func (cmw *cmuxWrapper) shutDown() error {
-	if cmw.grpcServer != nil {
-		cmw.grpcServer.GracefulStop()
-	}
-
-	if cmw.httpServer != nil {
-		shutdownCtx, cancel := context.WithTimeout(context.Background(), cmw.gracePeriod)
-		defer cancel()
+	// context has been cancelled (due to cancelled base-context or due to
+	// server exit).
+	shutdownCtx, cancel := context.WithTimeout(context.Background(), pmux.gracePeriod)
+	defer cancel()
 
-		return cmw.httpServer.Shutdown(shutdownCtx)
+	if pmux.grpcServer != nil {
+		pmux.grpcServer.GracefulStop()
 	}
 
-	return nil
+	return http1Server.Shutdown(shutdownCtx)
 }
 
-func serveOnListener(l net.Listener, server interface{ Serve(l net.Listener) error }, errCh chan<- error) {
-	if server == nil {
-		return
+func (pmux *protocolMuxServer) muxedHandler() http.Handler {
+	// if only one of gRPC and HTTP are set, no need to multiplex.
+	if pmux.grpcServer == nil {
+		return pmux.httpHandler
+	} else if pmux.httpHandler == nil {
+		return pmux.grpcServer
 	}
 
-	go func() {
-		err := server.Serve(l)
-		if err != nil {
-			if errors.Is(err, http.ErrServerClosed) || errors.Is(err, grpc.ErrServerStopped) {
-				return
-			}
-			errCh <- err
-			return
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.ProtoMajor == 2 && strings.Contains(r.Header.Get("Content-Type"), grpcRequestMarker) {
```

Comment:
will grpc insecure also use `ProtoMajor = 2` ?

Reply:
what do you mean "insecure" here? in context of http, "insecure usually refers to `do-not-verify-certificate-authority` (as in, when you pass `--insecure` to curl for example).. But i am assuming you are talking about plain HTTP (without the S)?

ProtoMajor refers to the major version of HTTP.. gRPC uses http2. i don't think it is impacted by HTTP vs HTTPS.